app/socket/service/activity-service.js

```diff
@@ -0,0 +1,143 @@
+const keys = require('../redis/keys');
+const utils = require('../utils');
+
+module.exports = (config, redis) => {
+  const ttl = {
+    user: config.get('redis.socket.userDetailsTtlSec'),
+    activity: config.get('redis.socket.activityTtlSec')
+  };
+
+  const notifyChange = (caseId) => {
+    if (caseId) {
+      redis.publish(keys.case.base(caseId), Date.now().toString());
+    }
+  };
+
+  const getSocketActivity = async (socketId) => {
+    if (socketId) {
+      const key = keys.socket(socketId);
+      return JSON.parse(await redis.get(key));
+    }
+    return null;
+  };
+
+  const getUserDetails = async (userIds) => {
+    if (Array.isArray(userIds) && userIds.length > 0) {
+      // Get hold of the details.
+      const details = await redis.pipeline(utils.get.users(userIds)).exec();
+
+      // Now turn them into a map.
+      return details.reduce((obj, item) => {
+        if (item[1]) {
+          const user = JSON.parse(item[1]);
+          obj[user.id] = { forename: user.forename, surname: user.surname };
+        }
+        return obj;
+      }, {});
+    }
+    return {};
+  };
+
+  const doRemoveSocketActivity = async (socketId) => {
+    // First make sure we actually have some activity to remove.
+    const activity = await getSocketActivity(socketId);
+    if (activity) {
+      await redis.pipeline([
+        utils.remove.userActivity(activity),
+        utils.remove.socketEntry(socketId)
+      ]).exec();
+      return activity.caseId;
+    }
+    return null;
+  };
+
+  const removeSocketActivity = async (socketId) => {
+    const removedCaseId = await doRemoveSocketActivity(socketId);
+    if (removedCaseId) {
+      notifyChange(removedCaseId);
+    }
+  };
+
+  const doAddActivity = async (caseId, user, socketId, activity) => {
```

Comment:
Can we have all the params caseId, user, socketId, activity strongly typed

Reply:
Given that it's JavaScript, not really...

Reply:
Ok